File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

long long solution(int n, vector<int> times) {
    sort(times.begin(), times.end());

    long long left = 1;
    long long right = (long long)times.back() * n;
    long long answer = right;

    while (left <= right) {
        long long mid = (left + right) / 2;
        long long sum = 0;

        for (int t : times) {
            sum += mid / t;
            if (sum >= n) break;
        }

        if (sum >= n) {
            answer = mid;
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    return answer;
}

File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public long solution(int n, int[] times) {
        Arrays.sort(times);
        
        long left = 1;
        long right = n * (long)times[times.length - 1];
        long answer = right;
        
        while (left <= right){
            long mid = (left + right) / 2;
            long sum = 0;
            
            for (int t : times){
                sum += mid / t;
                if (sum >= n) break;
            }
            
            // System.out.println(sum);
            
            if (sum >= n){
                right = mid - 1;
                answer = mid;
            } else {
                left = mid + 1;
            }
        }
        
        return answer;
    }
}